design-proposals/vm-memory-dump.md

```diff
@@ -0,0 +1,147 @@
+It will patch the vm status with a memoryDumpStatus stating the volume-name and Phase `Binding`.
+The memoryDumpStatus will look something like this:
+
+`
+// MemoryDumpStatus represent the memory dump request status and info
+type MemoryDumpStatus struct {
+	// VolumeName is the name of the pvc that will contain the memory dump
+	VolumeName string `json:"volumeName"`
+	// Phase represents the memory dump phase
+	Phase MemoryDumpPhase `json:"phase,omitempty"`
+	// TimeStamp represents the time the memory dump was completed
+	TimeStamp *metav1.Time `json:"timestamp,omitempty"`
+}
+
+type MemoryDumpPhase string
+
+const (
+	// The memorydump is during pvc binding
+	MemoryDumpBinding MemoryDumpPhase = "Binding"
+	// The memorydump is in progress
+	MemoryDumpInProgress MemoryDumpPhase = "InProgress"
+	// The memorydump is completed
+	MemoryDumpCompleted MemoryDumpPhase = "Completed"
+	// The memorydump is being unbound
+	MemoryDumpUnBinding MemoryDumpPhase = "Unbinding"
+)
+`
+
+In the virt controller once getting the update of the memoryDump in state of binding it will bind the pvc to the vm by updating the vm volumes with the memoryDump pvc.
+After that the status of the memoryDump will be updated to `InProgress` which will be cause the volume to be mounted to the virt-launcher, same as done in the hotplug process, other then the end of the process that instead of attaching the volume to the VMI it will trigger the guest memory dump `virDomainCoreDump` command with `VIR_DUMP_MEMORY_ONLY` flag to be executed by virt-launcher.
+(Check out [virDomainCoreDump](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainCoreDump) for reference)
+After the dump is complete the timestamp and status of the memory dump will be updated and it will unmount the pvc from virt-launcher.
+The PVC will remain bound to the VM - will be kept in the VM volumes and have the updated status and timestamp in the VM status as mentioned.
+It will look something like that: (In the VMI there will be no information)
+
+```yaml
+---
+apiVersion: v1
+items:
+- apiVersion: kubevirt.io/v1
+  kind: VirtualMachine
+  name: my-vm
+  ....
+  Snip
+  ....
+    template:
+      spec:
+        domain:
+          devices:
+            disks:
+            - disk:
+                bus: virtio
+              name: datavolumedisk
+            - disk:
+                bus: virtio
+              name: cloudinitdisk
+        volumes:
+        - dataVolume:
+            name: my-dv
+          name: datavolumedisk
+        - cloudInitNoCloud:
+            userData: |-
+              #cloud-config
+              password: fedora
+              chpasswd: { expire: False }
+          name: cloudinitdisk
+        - memoryDump:
+            persistentVolumeClaim:
+                claimName: memory-dump-pvc
+          name: memory-dump
+  status:
+    ....
+    Snip
+    ....
+    memoryDumpStatus:
```

Comment:
This is the last memory dump then?

Reply:
yep I mentioned it in the design, there will be information on the last dump, the user will be responsible to handle it before taking another dump that will ovewrite the previous if its on the same pvc